### PR TITLE
Add mode function

### PR DESCRIPTION
**Pull request by exfinen**
- Add mode function
- Remove code from test library that is no longer used due to removal of runtime_error from mp-spdz dsl based functions
- Remove unimplemented test helper function
- Remove magic numbers from the dataset and app an option to apply data filter upon testing correspinding Python function

### CODE REVIEW

File: mpcstats_lib.py [hunk 1]
```diff
@@ -181,18 +181,55 @@ def geometric_mean(data: list[sint]):
...
+        ))
+
+    # get the first elem in data with the highest frequency
+    highest = sint(0)
+
+    # reverse the data to select the first element with the highest
+    # frequency, aligning with the behavior of its python counterpart
+    rev_data = Array.create_from(data.get_reverse_vector())
+
+    @for_range(n)
```
Comment (mhchia) [inline mpcstats_lib.py:217]: I'm thinking we can just range `i` from `n-1` to `0` so we don't need to create another vector `rev_data`. Would this make sense?
```suggestion
    @for_range(n-1, -1, -1)
```

**Review by mhchia — COMMENTED** [review]
This implementation looks good to me, thanks! I just found that MP-SPDZ has an [Array.sort](https://mp-spdz.readthedocs.io/en/latest/Compiler.html#Compiler.types.Array.sort) function and wondered if it could help with performance. By sorting the list first, we can find the most common element in a single iteration.

Here's an example:
```python
data = Array.create_from(data)
data.sort()
most_common = data[0]
most_common_count = sint(1)
largest_count = sint(1)
for index, ele in enumerate(data[1:]):
    most_common_count.update(if_else(ele == data[index], most_common_count + 1, sint(1)))
    largest_count.update(if_else(most_common_count > largest_count, most_common_count, largest_count))
    most_common.update(if_else(most_common_count > largest_count, ele, most_common))
```

According to the documentation, sorting is O(nlogn), so this approach should be asymptotically faster. What do you think? Would this make sense, or might there be any issues?

**Comment by exfinen** [comment]
> This implementation looks good to me, thanks! I just found that MP-SPDZ has an [Array.sort](https://mp-spdz.readthedocs.io/en/latest/Compiler.html#Compiler.types.Array.sort) function and wondered if it could help with performance. By sorting the list first, we can find the most common element in a single iteration.
> 
> Here's an example:
> 
> ```python
> data = Array.create_from(data)
> data.sort()
> most_common = data[0]
> most_common_count = sint(1)
> largest_count = sint(1)
> for index, ele in enumerate(data[1:]):
>     most_common_count.update(if_else(ele == data[index], most_common_count + 1, sint(1)))
>     largest_count.update(if_else(most_common_count > largest_count, most_common_count, largest_count))
>     most_common.update(if_else(most_common_count > largest_count, ele, most_common))
>     ...
> ```
> 
> According to the documentation, sorting is O(nlogn), so this approach should be asymptotically faster. What do you think? Would this make sense, or might there be any issues?

I have also tried the idea, but was unable to find a way to only scan the data array once.
In the example, because you scan the data array in the inner loop, I think the code boils down to:

```python
sort in O(n log n)

for 2 to n {
  for 2 to n { do_something }
  for 2 to n { do_something }
  for 2 to n { do_something }
}
```
which is O(n log n) + O(n^2) = O(n^2). Do you think this interpretation is correct?

File: mpcstats_lib.py (hunk 1 above)
Comment (exfinen) [reply mpcstats_lib.py]: Yes. I wasn't aware that such indexing was possible. Fixed the code.

**Comment by exfinen** [comment]
Please forget about my previous comment which was misunderstanding the inner loop :)

I now realize that the time copmlexity of the example is O(n log n) + O(n) = O(n log n).
Whereas the current code has 4 * O(n) = O(n) time complexity, so I think it's faster for large enough inputs. 
Do you agree? Maybe I'm missing something.

**Comment by mhchia** [comment]
Ah I think a for here should be O(n^2). Because `sum` goes through `data` (n elements) and `for_range(n)` is n elements, together it's O(n*n) = O(n^2).
```python
    @for_range(n)
    def _(i):
        freqs[i] = sum(if_else(x == data[i], 1, 0) for x in data)
```

**Comment by exfinen** [comment]
Sorry. It seems stupid, but I missed that part! I will rewrite ir using the sort scheme.

**Comment by mhchia** [comment]
No worries I thought it were O(n) at first as well 😂

**Comment by exfinen** [comment]
I needed slight modificatoin and it's working now. 😃  I am now wondering  how i should handle the magic numbers.

**Comment by exfinen** [comment]
Perhaps, I initially set the value to return to the magic number and don't update the most frequent element if the element is the magic number. I will try it when I get home.

**Comment by exfinen** [comment]
Another idea is that first find the range of magic number indices and scan the range in two passes

**Comment by exfinen** [comment]
Setting count to zero if the current number is magic number. Now a new problem is that because the list is sorted, the result can be different from the result of Python function. 

For example, when the dataset is `[1, 3, 3, 2, 2]`, Python mode function returns 3 if , but our current implementation returns 2. @mhchia is this difference acceptable?

**Comment by mhchia** [comment]
That's a good catch! in this case, I think we should conform to the python lib and it might be simpler to use your original implementation. What's your thoughts?

**Comment by gogoex** [comment]
So keeping the same behavior as Python stat function is important. Then I think it makes sense to use the original one. We may be able to improve the running time in the future if we come up with a better idea.

I reverted the change with a small improvement to avoid finding frequency for magic numbers.
It passes all the newly added test cases.

**Comment by exfinen** [comment]
So keeping the same behavior as Python stat function is important. Then I think it makes sense to use the original one. We may be able to improve the running time in the future if we come up with a better idea.

I reverted the change with a small improvement to avoid finding frequency for magic numbers.
It passes all the newly added test cases.

**Comment by exfinen** [comment]
I also wondered what happens if multiply the frequency by a weight that is len(data) - i. But this way, earlier value can win the actual highest.

**Review by mhchia — APPROVED** [review]
Looks good to me :) Thanks

**Pull request merged**